File: design/src/main/java/com/kimjisub/design/Panel/StorePackPanel.java
package com.kimjisub.design.panel;

import android.content.Context;
import android.content.res.TypedArray;
import android.util.AttributeSet;
import android.view.LayoutInflater;
import android.view.View;
import android.widget.RelativeLayout;

import androidx.databinding.DataBindingUtil;

import com.kimjisub.design.R;
import com.kimjisub.design.databinding.PanelStorePackBinding;

public class StorePackPanel extends RelativeLayout {

	PanelStorePackBinding b;

	public StorePackPanel(Context context) {
		super(context);
		init();
	}

	public StorePackPanel(Context context, AttributeSet attrs) {
		super(context, attrs);
		init();
		getAttrs(attrs);
	}

	public StorePackPanel(Context context, AttributeSet attrs, int defStyleAttr) {
		super(context, attrs, defStyleAttr);
		init();
		getAttrs(attrs, defStyleAttr);
	}

	private void init() {
		b = DataBindingUtil.inflate(LayoutInflater.from(getContext()), R.layout.panel_store_pack, this, true);

		b.title.setSelected(true);
		b.subTitle.setSelected(true);
		b.path.setSelected(true);

		b.youtube.setOnClickListener(v -> {
			if (onEventListener != null)
				onEventListener.onYoutubeClick(v);
		});
		b.website.setOnClickListener(v -> {
			if (onEventListener != null)
				onEventListener.onWebsiteClick(v);
		});
	}

	private void getAttrs(AttributeSet attrs) {
		TypedArray typedArray = getContext().obtainStyledAttributes(attrs, R.styleable.MainPackPanel);
		setTypeArray(typedArray);
	}

	private void getAttrs(AttributeSet attrs, int defStyle) {
		TypedArray typedArray = getContext().obtainStyledAttributes(attrs, R.styleable.MainPackPanel, defStyle, 0);
		setTypeArray(typedArray);
	}

	private void setTypeArray(TypedArray typedArray) {

		typedArray.recycle();
	}

	////////////////////////////////////////////////////////////////////////////////////////////////

	public void setTitle(String title) {
		b.title.setText(title);
	}

	public void setSubTitle(String subTitle) {
		b.subTitle.setText(subTitle);
	}

	public void setDownloadCount(String downloadCount) {
		b.downloadCount.setText(downloadCount);
	}

	public void updateTitle(String title) {
		if (!b.title.getText().equals(title)) {
			b.title.setAlpha(0);
			setTitle(title);
			b.title.animate().alpha(1).setDuration(500).start();
		}
	}

	public void updateSubTitle(String subTitle) {
		if (!b.subTitle.getText().equals(subTitle)) {
			b.subTitle.setAlpha(0);
			setSubTitle(subTitle);
			b.subTitle.animate().alphaBy(0).alpha(1).setDuration(500).start();
		}
	}

	public void updateDownloadCount(String downloadCount) {
		if (!b.downloadCount.getText().equals(downloadCount)) {
			b.downloadCount.setAlpha(0);
			setDownloadCount(downloadCount);
			b.downloadCount.animate().alphaBy(0).alpha(1).setDuration(500).start();
		}
	}

	////////////////////////////////////////////////////////////////////////////////////////////////

	OnEventListener onEventListener;

	public interface OnEventListener {

		void onYoutubeClick(View v);

		void onWebsiteClick(View v);
	}

	public void setOnEventListener(OnEventListener onEventListener) {
		this.onEventListener = onEventListener;
	}
}

File: design/src/main/java/com/kimjisub/design/panel/StorePackPanel.kt
package com.kimjisub.design.panel

import android.content.Context
import android.util.AttributeSet
import android.view.LayoutInflater
import android.view.View
import android.widget.RelativeLayout
import androidx.databinding.ObservableField
import com.kimjisub.design.R.styleable
import com.kimjisub.design.databinding.PanelStorePackBinding
import java.text.DecimalFormat

class StorePackPanel
@JvmOverloads constructor(
	context: Context,
	attrs: AttributeSet? = null,
	defStyleAttr: Int = 0,
) : RelativeLayout(context, attrs, defStyleAttr) {
	val b: PanelStorePackBinding =
		PanelStorePackBinding.inflate(LayoutInflater.from(context), this, true)
	val data = Data()

	class Data {
		val title: ObservableField<String> = ObservableField()
		val subtitle: ObservableField<String> = ObservableField()
		val downloadCount: ObservableField<String> = ObservableField()
		val path: ObservableField<String> = ObservableField()
	}

	init {
		attrs?.let {
			val typedArray =
				context.obtainStyledAttributes(it, styleable.StorePackPanel, defStyleAttr, 0)

			//	b.TVTitle.isSelected = true
			//	b.TVSubtitle.isSelected = true
			//	b.path.isSelected = true
			b.youtube.setOnClickListener { v: View ->
				onEventListener?.onYoutubeClick(v)
			}
			b.website.setOnClickListener { v: View ->
				onEventListener?.onWebsiteClick(v)
			}

			typedArray.recycle()
		}
	}

	private val numberFormatter = DecimalFormat("###,###")

	////////////////////////////////////////////////////////////////////////////////////////////////


	fun setTitle(title: String) {
		b.title.text = title
	}

	fun setSubtitle(subtitle: String) {
		b.subtitle.text = subtitle
	}

	fun setDownloadCount(downloadCount: String) {
		b.downloadCount.text = downloadCount
	}

	fun setDownloadCount(downloadCount: Long) {
		val downloadCountFormatted: String = numberFormatter.format(downloadCount)
		setDownloadCount(downloadCountFormatted)
	}

	fun updateTitle(title: String) {
		if (b.title.text != title) {
			b.title.alpha = 0f
			setTitle(title)
			b.title.animate().alpha(1f).setDuration(500).start()
		}
	}

	fun updateSubtitle(subtitle: String) {
		if (b.subtitle.text != subtitle) {
			b.subtitle.alpha = 0f
			setSubtitle(subtitle)
			b.subtitle.animate().alphaBy(0f).alpha(1f).setDuration(500).start()
		}
	}

	fun updateDownloadCount(downloadCount: Long) {
		val downloadCountFormatted: String = numberFormatter.format(downloadCount)
		if (b.downloadCount.text != downloadCountFormatted) {
			b.downloadCount.alpha = 0f
			setDownloadCount(downloadCountFormatted)
			b.downloadCount.animate().alphaBy(0f).alpha(1f).setDuration(500).start()
		}
	}

	////////////////////////////////////////////////////////////////////////////////////////////////


	var onEventListener: OnEventListener? = null

	interface OnEventListener {
		fun onYoutubeClick(v: View)
		fun onWebsiteClick(v: View)
	}
}